File: hash/go/imagehash/dhash.go
package imagehash

import (
	"fmt"
	"image"

	"github.com/wmetcalf/rosetta-image-hash/go/imagehash/internal/imgrgb"
	"github.com/wmetcalf/rosetta-image-hash/go/imagehash/internal/lanczos"
)

// DHash computes dhash: grayscale → Lanczos resize to (W=N+1, H=N) →
// row-wise adjacent-column diff with strict >.
func DHash(img image.Image, hashSize int) (Hash, error) {
	if hashSize < 2 {
		return Hash{}, fmt.Errorf("hashSize must be >= 2, got %d", hashSize)
	}
	rgb := imgrgb.ToRGB(img)
	gray := rgbToGray(rgb)
	resized := lanczos.Resize(gray, hashSize+1, hashSize)
	bits := make([][]bool, hashSize)
	for y := 0; y < hashSize; y++ {
		bits[y] = make([]bool, hashSize)
		for x := 0; x < hashSize; x++ {
			bits[y][x] = resized[y][x+1] > resized[y][x]
		}
	}
	return newHashFromBits(bits), nil
}

File: hash/swift/RosettaSquintHash/Sources/RosettaSquintHash/DHash.swift
import Foundation

/// dhash: grayscale → Lanczos to (W=N+1, H=N) → row-wise adjacent-column diff (strict >).
public func dhash(_ image: RGBImage, hashSize: Int) throws -> Hash {
    try image.validate()
    guard hashSize >= 2 else { throw ImageHashError.invalidHashSize(hashSize) }
    let rgb = ImgRGB.toRGB(image)
    let gray = rgbToGray(rgb.data, width: rgb.width, height: rgb.height)
    let resized = lanczosResize(gray, srcW: rgb.width, srcH: rgb.height, dstW: hashSize + 1, dstH: hashSize)

    var bits: [[Bool]] = []
    for y in 0..<hashSize {
        let rowOff = y * (hashSize + 1)
        var row = [Bool](repeating: false, count: hashSize)
        for x in 0..<hashSize {
            row[x] = resized[rowOff + x + 1] > resized[rowOff + x]
        }
        bits.append(row)
    }
    return try Hash(bits: bits)
}
